transaction/builder/src/transaction_builder.rs

```diff
@@ -4129,4 +4130,361 @@ pub mod transaction_builder_tests {
+                .unwrap();
+
+            let TxOutContext {
+                tx_out: decoy_tx_out,
+                ..
+            } = transaction_builder
+                .add_change_output(Amount::new(0, token_id), &change_destination, &mut rng)
+                .unwrap();
+
+            let tx = transaction_builder
+                .build(&NoKeysRingSigner {}, &mut rng)
+                .expect("build tx");
+
+            assert_eq!(tx.prefix.outputs.len(), 2);
+
+            let main_output = tx
+                .prefix
+                .outputs
+                .iter()
+                .find(|tx_out| tx_out.public_key == defrag_tx_out.public_key)
+                .expect("Didn't find recipient's output");
+            let decoy_output = tx
+                .prefix
+                .outputs
+                .iter()
+                .find(|tx_out| tx_out.public_key == decoy_tx_out.public_key)
+                .expect("Didn't find sender's output");
+
+            // Defrag output should have a defrag memo
+            let ss = get_tx_out_shared_secret(
+                sender.view_private_key(),
+                &RistrettoPublic::try_from(&main_output.public_key).unwrap(),
+            );
+            let memo = main_output.e_memo.unwrap().decrypt(&ss);
+            match MemoType::try_from(&memo).expect("Couldn't decrypt memo") {
+                MemoType::Defragmentation(memo) => {
+                    assert_eq!(memo.get_defrag_id(), 0u64);
+                    assert_eq!(memo.get_fee(), 3u64);
+                    assert_eq!(memo.get_total_outlay(), 432u64);
+                }
+                _ => {
+                    panic!("unexpected memo type")
+                }
+            }
+
+            // Change output should have a 0 value defrag memo
+            let ss = get_tx_out_shared_secret(
+                sender.view_private_key(),
+                &RistrettoPublic::try_from(&decoy_output.public_key).unwrap(),
+            );
+            let memo = decoy_output.e_memo.unwrap().decrypt(&ss);
+            match MemoType::try_from(&memo).expect("Couldn't decrypt memo") {
+                MemoType::Defragmentation(memo) => {
+                    assert_eq!(memo.get_defrag_id(), 0u64);
+                    assert_eq!(memo.get_fee(), 0u64);
+                    assert_eq!(memo.get_total_outlay(), 0u64);
+                }
+                _ => {
+                    panic!("unexpected memo type")
+                }
+            }
+        }
+
+        // Test builds memos with ID
+        {
+            let memo_builder = DefragmentationMemoBuilder::new(64);
+
+            let mut transaction_builder = TransactionBuilder::new(
+                block_version,
+                Amount::new(10, token_id),
+                fog_resolver.clone(),
+                memo_builder,
+            )
+            .unwrap();
+
+            transaction_builder.set_fee(3).unwrap();
+
```

Comment:
:wrench: similar to above, fee seems unecessary
```suggestion
```

Reply:
Same as above.